### DISCUSSION TITLE
lesson 2 mapping : how does mapping into the same id work

### DISCUSSION THREAD

**Discussion by note-xin** [Cyfrin/foundry-full-course-cu#1364] [Q&A] [answered]
I have mapped a person with name = 'tom' to fav_number = 3, and later mapped another person with the same name to a fav_number = 333. When I enquired about 'tom' the fav_number was 333 and not 3. How does it work?

**Comment by ghost** [comment, marked as answer]
Mappings in Solidity are just hash tables. A hash table is a data structure that implements an associative array, I think also called a "dictionary". As show in the image below, in a associative array, data is stored in a collection of key-value pairs.

![image](https://github.com/Cyfrin/foundry-full-course-f23/assets/109909308/4c35720f-b0b4-427b-91cb-7f455219cbe4)

To determine in which position the data will be stored, the data (or key) is passed to a hash function which produces a hashed value that indicates the index within the hash table where the data will be stored.

When you want to lookup for a vaue, all you need to do is pass the key, the key is hashed, the position is determined and the data is fetched. This is why is called a dictionary, because it works like that, a key always maps to the same position.

So when you do something like `mapping(string => uint256)` you are saying:

> Solidity, create a table view that maps a `string` type to a `uint256` type that will help me organize my data so I can easliy store it, update it and retrieve it.

Due to the nature of hash functions **a key will always produce the same hashed value**, in other words, the string "Tom" always access the same slot in storage (to store, update and read).

Let's assume that "Tom" maps to slot 10,  when you pass "Tom" as a key and 3 as the data to store that value will be stored in the slot 10, and when you read from the mapping by passing the string "Tom", it will always read from the slot 10. In the same way when you pass "Tom" and 333, the content of the slot will be replaced with the new value.

**Reply by note-xin** [reply]
Thank you for taking your valuable time to clear my doubts. Much appreciated!
- peace be upon you.
